### PR TITLE
[forge] nodegroup resize

### PR DESCRIPTION
**Pull request by rustielin**
## Motivation

Add the ability for Forge to resize its k8s backend clusters' worker nodegroups. Before, we could only uninstall helm releases, but would always have the worker nodes running. This saves us a lot of $$$.

Uses the `rusoto` and `rusoto_eks` crates to interact with EKS and make the API calls. An additional dep `hyper-proxy` is introduced to make things work with an http proxy.

Refactoring:
* `remove_validator`/`remove_testnet` now use `remove_helm_release`
* `k8s_client` -> `create_k8s_client`
* Separate `clean-up` subcommand for forge CLI into `clean-up` (to 0 validators) and `resize` (reset to N validators)

NOTE: for now, the cluster does not scale down/go into idle state after test run, as that logic seems to have been removed in a previous PR.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

Run the operator CLI:

```
$ time cargo run -p forge-cli -- --cluster-name libra-rustietest --helm-repo testnet-rustietest clean-up
<helm uninstalls>
All validators removed
Created rusoto http client
Created validators nodegroup update request with ID: 8a1ae8f2-b828-33c9-a89c-a4b2576692c4
Created utilities nodegroup update request with ID: 4b14b3b4-d198-3e4f-becd-744ee0a2b805
Created trusted nodegroup update request with ID: 9e8d88a1-9f47-3088-a49c-3e712cf1d13d
Waiting for utilities update 4b14b3b4-d198-3e4f-becd-744ee0a2b805: InProgress ...
Waiting for validators update 8a1ae8f2-b828-33c9-a89c-a4b2576692c4: InProgress ...
Waiting for trusted update 9e8d88a1-9f47-3088-a49c-3e712cf1d13d: InProgress ...
<waiting>
validators nodegroup update 8a1ae8f2-b828-33c9-a89c-a4b2576692c4 successful!!!
utilities nodegroup update 4b14b3b4-d198-3e4f-becd-744ee0a2b805 successful!!!
trusted nodegroup update 9e8d88a1-9f47-3088-a49c-3e712cf1d13d successful!!!
27.67user 3.90system 1:10.04elapsed 45%CPU (0avgtext+0avgdata 60040maxresident)k
32inputs+0outputs (0major+817392minor)pagefaults 0swaps
...

$ cargo run -p forge-cli -- --cluster-name libra-rustietest --helm-repo testnet-rustietest resize --num-validators 4
<same as before>
```

Also from `fgi`

```
# reset the cluster to idle state
$ time cargo run -p forge-cli -- --cluster-name libra-rustietest --helm-repo testnet-rustietest clean-up
<same as above>

# compat test completes within 15 min from idle state!
# full output P445974160
$ time ./scripts/fgi/run -T dev_rustielin_pull_8982 -W rustietest
...
test result: ok. 5 passed; 0 failed; 0 filtered out

==========end-pod-logs==========

**********
Dashboard snapshot: http://mon.rustietest.aws.hlw3truzy4ls.com/d/overview/overview?from=1629335729182&to=1629336555283&orgId=1
**********

Job forge-rustielin-1629335716 succeeded!

real    14m2.617s
user    0m11.787s
sys     0m2.332s
```

### CODE REVIEW

File: testsuite/forge-cli/src/main.rs [hunk 1]
```diff
@@ -105,14 +116,25 @@ fn main() -> Result<()> {
...
                 )
             }
-            OperatorCommand::CleanUp(cleanup) => {
+            OperatorCommand::Resize(resize) => {
+                let _ = set_eks_nodegroup_size(
+                    args.cluster_name,
+                    resize.num_validators,
+                    resize.auth_with_k8s_env,
+                )
+                .unwrap();
```
Comment (bmwill) [inline testsuite/forge-cli/src/main.rs:125]: There's lots of places in this pr where you assign the return value of a function which is a `Result` to `_`. This ignores the error. Instead we probably want to propagate the error up like this:
 
```suggestion
                    set_eks_nodegroup_size(
                    args.cluster_name,
                    resize.num_validators,
                    resize.auth_with_k8s_env,
                )?;
```

File: testsuite/forge-cli/src/main.rs [hunk 2]
```diff
@@ -105,14 +116,25 @@ fn main() -> Result<()> {
...
+                let _ = uninstall_from_k8s_cluster().unwrap();
```
Comment (bmwill) [inline testsuite/forge-cli/src/main.rs:126]: Same here

File: testsuite/forge-cli/src/main.rs [hunk 3]
```diff
@@ -105,14 +116,25 @@ fn main() -> Result<()> {
...
-                    cleanup.require_validator_healthcheck,
-                )
+                    resize.num_validators,
+                    resize.validator_image_tag,
+                    resize.testnet_image_tag,
+                    resize.require_validator_healthcheck,
+                );
+            }
+            OperatorCommand::CleanUp(cleanup) => {
+                let _ = uninstall_from_k8s_cluster().unwrap();
```
Comment (bmwill) [inline testsuite/forge-cli/src/main.rs:136]: here

File: testsuite/forge/src/backend/k8s/mod.rs [hunk 4]
```diff
@@ -86,6 +93,8 @@ impl Factory for K8sFactory {
         node_num: NonZeroUsize,
         version: &Version,
     ) -> Result<Box<dyn Swarm>> {
+        let _ = set_eks_nodegroup_size(self.cluster_name.clone(), node_num.get(), true);
+        let _ = uninstall_from_k8s_cluster();
         let _ = clean_k8s_cluster(
```
Comment (bmwill) [inline testsuite/forge/src/backend/k8s/mod.rs:98]: these three lines as well

**Review by zihaoccc — COMMENTED** [review]
generally looks good to me, awesome work.

File: testsuite/forge/src/backend/k8s/cluster_helper.rs [hunk 5]
```diff
@@ -310,3 +342,183 @@ pub async fn k8s_client() -> K8sClient {
...
+    // submit the scaling requests
+    let rt = Runtime::new().unwrap();
+    let validators_update_id = rt
+        .block_on(submit_update_nodegroup_config_request(
+            &eks_client,
+            &cluster_name,
+            "validators",
+            validator_scaling,
+        ))
+        .unwrap();
```
Comment (bmwill) [inline testsuite/forge/src/backend/k8s/cluster_helper.rs:451]: Lets return the error instead of panicking
```suggestion
        ?;
```

Comment (bmwill) [reply testsuite/forge/src/backend/k8s/cluster_helper.rs:451]: There's also a couple other places in this diff that should be tweaked too.

File: testsuite/forge/src/backend/k8s/cluster_helper.rs [hunk 6]
```diff
@@ -61,84 +72,86 @@ pub fn set_validator_image_tag(
...
+fn upgrade_validator(validator_name: &str, helm_repo: &str, options: &[&str]) -> Result<()> {
+    upgrade_helm_release(
+        validator_name,
+        &format!("{}/diem-validator", helm_repo),
+        options,
+    )
+}
 
-    Ok(())
+fn upgrade_testnet(helm_repo: &str, options: &[&str]) -> Result<()> {
```
Comment (bmwill) [inline testsuite/forge/src/backend/k8s/cluster_helper.rs:153]: I think i'd like for the options to be a bit more strongly typed than just a slice of strings but given this is an internal to this file api it's fine as is.

**Review by bmwill — COMMENTED** [review]
There's a few more places where we could change `unwrap -> ?` and then we can get this landed!

File: testsuite/forge/src/backend/k8s/cluster_helper.rs [hunk 7]
```diff
@@ -310,3 +342,183 @@ pub async fn k8s_client() -> K8sClient {
...
+                    };
+                    Box::pin(async move {
+                        let describe_update = client
+                            .describe_update(describe_update_request)
+                            .await
+                            .unwrap()
+                            .update
+                            .unwrap();
+                        let stat = describe_update.status;
+                        match stat {
```
Comment (zihaoccc) [inline testsuite/forge/src/backend/k8s/cluster_helper.rs:496]: I reckon `if let Some()` here can make this piece more readable and clearer

**Review by zihaoccc — COMMENTED** [review]
Overall LGTM, super close to land. please address last few comments and let's land it.
Thx for your effort to save $$$$ 😆

**Review by bmwill — APPROVED** [review]

**Review by bmwill — DISMISSED** [review]

**Pull request merged**